metatensor-torch/include/metatensor/torch/atomistic.hpp

```diff
@@ -0,0 +1,388 @@
+    ///        these vectors (i.e. the cell should be given in row-major order).
+    ///        Systems are assumed to obey periodic boundary conditions,
+    ///        non-periodic systems should set the cell to 0.
+    SystemHolder(torch::Tensor species, torch::Tensor positions, torch::Tensor cell);
+    ~SystemHolder() override = default;
+
+    /// Get the species for all particles in the system.
+    torch::Tensor species() const {
+        return species_;
+    }
+
+    /// Get the positions for the atoms in the system.
+    torch::Tensor positions() const {
+        return positions_;
+    }
+
+    /// Unit cell/bounding box of the system.
+    torch::Tensor cell() const {
+        return cell_;
+    }
+
+    /// Get the device used by all the data in this `System`
+    torch::Device device() const {
+        return species_.device();
+    }
+
+    /// Get the dtype used by all the floating point data in this `System`
+    torch::Dtype scalar_type() const {
+        return positions_.scalar_type();
+    }
+
+    /// Get the number of particles in this system
+    int64_t size() const {
+        return  this->species_.size(0);
+    }
+
+    // TODO: add `SystemHolder::to(dtype, device) -> TorchSystem` to convert
+    // dtype & device for a System
+
+    /// Add a new neighbors list in this system corresponding to the given
+    /// `options`.
+    ///
+    /// The neighbors list should have the following samples: "first_atom",
+    /// "second_atom", "cell_shift_a", "cell_shift_b", "cell_shift_c",
+    /// containing the index of the first and second atom and the number of
+    /// cell vector a/b/c to add to the positions difference to get the pair
+    /// vector.
+    ///
+    /// The neighbors should also have a single component "xyz" with values `[0,
+    /// 1, 2]`; and a single property "distance" with value 0.
+    ///
+    /// The neighbors values must contain the distance vector from the first to
+    /// the second atom, i.e. `positions[second_atom] - positions[first_atom] +
+    /// cell_shift_a * cell_a + cell_shift_b * cell_b + cell_shift_c * cell_c`.
+    void add_neighbors_list(NeighborsListOptions options, TorchTensorBlock neighbors);
+
+    /// Retrieve a previously stored neighbors list with the given options, or
+    /// throw an error if no such neighbors list exists.
+    TorchTensorBlock get_neighbors_list(NeighborsListOptions options) const;
+
+    /// Get the list of neighbors lists registered with this `System`
+    std::vector<NeighborsListOptions> known_neighbors_lists() const;
+
+    /// Add custom data to this system, stored as `TensorBlock`.
+    ///
+    /// This is intended for experimentation with models that need more data as
+    /// input, and moved into a field of `SystemHolder` later.
+    ///
+    /// @param name name of the data
+    /// @param values values of the data
+    void add_data(std::string name, TorchTensorBlock values);
```

Comment:
Do we have particular types of data in mind here? I would do `TensorMap`s to have more flexibility (e.g., two quantities that include components of different symmetries (L) each)